src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -513,23 +514,24 @@ public int compare(DiscoveryNode o1, DiscoveryNode o2) {
         return changed;
     }
 
-    static class InternalAsyncFetch<T extends BaseNodeResponse> extends AsyncShardFetch<T> {
+    private final AtomicBoolean rerouting = new AtomicBoolean();
 
-        private final ClusterService clusterService;
-        private final AllocationService allocationService;
+    class InternalAsyncFetch<T extends BaseNodeResponse> extends AsyncShardFetch<T> {
 
-        public InternalAsyncFetch(ESLogger logger, String type, ShardId shardId, List<? extends BaseNodesResponse<T>, T> action,
-                                  ClusterService clusterService, AllocationService allocationService) {
+        public InternalAsyncFetch(ESLogger logger, String type, ShardId shardId, List<? extends BaseNodesResponse<T>, T> action) {
             super(logger, type, shardId, action);
-            this.clusterService = clusterService;
-            this.allocationService = allocationService;
         }
 
         @Override
         protected void reroute(ShardId shardId, String reason) {
-            clusterService.submitStateUpdateTask("async_shard_fetch(" + type + ") " + shardId + ", reasons (" + reason + ")", Priority.HIGH, new ClusterStateUpdateTask() {
+            if (rerouting.compareAndSet(false, true) == false) {
+                logger.trace("{} already has pending reroute, ignoring {}", shardId, reason);
+                return;
+            }
+            clusterService.submitStateUpdateTask("async_shard_fetch", Priority.HIGH, new ClusterStateUpdateTask() {
```

Comment:
my thought that it becomes less relevant, since a single reroute actually represents a few events now, and it can be misleading seeing in the pending task information about a shard id, where it might be ones for multiple ones